lib/Controller/SAMLController.php

```diff
@@ -309,31 +309,46 @@ public function assertionConsumerService() {
 	}
 
 	/**
+	 * @PublicPage
 	 * @NoAdminRequired
 	 * @NoCSRFRequired
 	 *
 	 * @return Http\RedirectResponse
 	 * @throws Error
 	 */
 	public function singleLogoutService() {
-
-		$pass = $this->request->passesCSRFCheck();
-		$isGlobalScaleEnabled = $this->config->getSystemValue('gs.enabled', false);
-		$gssMode = $this->config->getSystemValue('gss.mode', '');
-		if (!$pass && $isGlobalScaleEnabled && $gssMode === 'master') {
+		$targetUrl = null;
+		$isFromGS = ($this->config->getSystemValue('gs.enabled', false) &&
+					 $this->config->getSystemValue('gss.mode', '') === 'master');
+		$isFromIDP = !$isFromGS && !empty($_GET['SAMLRequest']);
+
+		if($isFromIDP) {
+			// requests comes from the IDP so let it manage the logout
+			// (or raise Error if request is invalid)
+			$pass = True ;
+		} elseif($isFromGS) {
+			// Request is from master GlobalScale
+			// Request validity is check via a JSON Web Token
 			$jwt = $this->request->getParam('jwt', '');
 			$pass = $this->isValidJwt($jwt);
+		} else {
+			// standard request : need read CRSF check
+			$pass = $this->request->passesCSRFCheck();
 		}
 
-		if($pass) {
+		if($pass){
 			$idp = $this->session->get('user_saml.Idp');
 			$auth = new Auth($this->SAMLSettings->getOneLoginSettingsArray($idp));
-			$returnTo = null;
+			$this->userSession->logout();
+			if($isFromIDP){
+				// processSLO will do the redirection himself
+				return $auth->processSLO();
+			}
+			// If request is not from IDP, we must send him the logout request
 			$parameters = array();
 			$nameId = $this->session->get('user_saml.samlNameId');
 			$sessionIndex = $this->session->get('user_saml.samlSessionIndex');
-			$this->userSession->logout();
-			$targetUrl = $auth->logout($returnTo, $parameters, $nameId, $sessionIndex, true);
+			$targetUrl = $auth->logout(null, $parameters, $nameId, $sessionIndex, true);
```

Comment:
while at it, you can also replace `$parameters` with a simple `[]`

Reply:
done :wink: